nixos/rke2: add rke2 service

mogeko opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

I added Systemd service to RKE2 (Rancher's next-generation Kubernetes distribution)!

RKE2 draws on a lot of K3S features in design, but it more closely aligned with upstream Kubernetes.  Therefore, I also use [K3S Service](https://github.com/NixOS/nixpkgs/blob/6afb255d976f85f3359e4929abd6f5149c323a02/nixos/modules/services/cluster/k3s/default.nix) as a reference when writing this Service.


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

I added RKE2 service and I have designed some very practical options.

You can easily select CNI, set disable services, etc.

```nix
services.rke2.enable = true;
services.rke2.cni = "cilium";
services.rke2.nodeName = "nixos-rke2-node0"；
services.rke2.disable = [
  "rke2-ingress-nginx"
];
# ...
```

This module will automatically add configuration files to [resolve conflicts between the default CNI and `NetworkManger`](https://docs.rke2.io/known_issues#networkmanager).

At this stage, I have only tested this module on my personal computer. In my future work, I will:

- [x] Complete the document
- [x] Add a test cases

---

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rorosen commented on nixos/modules/services/cluster/rke2/default.nix:
```diff
@@ -0,0 +1,276 @@
+
+    disable = mkOption {
+      type = with types;
+        listOf (enum [ "rke2-coredns" "rke2-ingress-nginx" "rke2-metrics-server" ]);
```
If this works similar to the disable option in k3s, which it does according to the [docs](https://docs.rke2.io/helm#disabling-addons), you can disable not only built-in add-ons but also any arbitrary manifest from the manifest directory. This shouldn't be an enum but allow any non-empty string.

mogeko replied:
I don't rule out that more will be added in the future, but according to this [docs](https://docs.rke2.io/reference/server_config#components) (and my experiment), the default startup of AddOns are just `rke2-coredns`, `rke2-ingress-nginx` and `rke2-metrics-server`.

rorosen replied:
Yes, that's true. However, if you place a file `my-addon.yaml` in `/var/lib/rancher/rke2/server/manifests/`, the add-on controller will apply every resource in the file. If you now want to let the add-on controller delete the applied resources, you need to pass `--disable my-addon`. This wouldn't be possible with the `disable` option, if it takes an enum.

I guess the important part of the [docs](https://docs.rke2.io/helm#disabling-addons) is 
> in addition to AddOns for any additional manifests placed in the manifests directory

superherointj commented:
I suggest you add some tests. For inspiration again:
https://github.com/NixOS/nixpkgs/tree/master/nixos/tests/k3s

Also, keep commits squashed. A single commit per NixOS module (nixos/rke2: xxx), NixOS test (nixos/tests/rke2: xxx) is ideal. Unless you are doing something unusual. But as you adding it new, it should be a single commit.

Also add test to the rke2 package, such that for whatever change to package it gets triggered and tested.

superherointj commented:
> Also, keep commits squashed. A single commit per NixOS module (nixos/rke2: xxx), NixOS test (nixos/tests/rke2: xxx) is ideal. Unless you are doing something unusual. But as you are adding it new, it should be a single commit.
> 
> Also add the test to the rke2 package (rke2: xxx), such that for whatever change to package it gets triggered and tested.

On commit messages, it should be:
* Package: `rke2: xxx`
* NixOS module:  `nixos/rke2: xxx`
* NixOS test: `nixos/tests/rke2: xxx`
Also, please squash commits.

Hope it makes sense.

mogeko commented:
I believe that this PR is ready to review and merge!

superherointj commented:
Please squash commits that belong to the same: package, nixos module, nixos test.

superherointj commented:
Commit: [nixos/rke2: complete the document](https://github.com/NixOS/nixpkgs/pull/308090/commits/819fe7eee572b319d457cd44cfd323e98f0d820c)
Has CI failing. Please fix error.
https://github.com/NixOS/nixpkgs/actions/runs/8910965374/job/24471302839

mogeko committed:
rke2: add runtime dependencies for kubelet

See: https://github.com/kubernetes/kubernetes/issues/26093#issuecomment-237202494

Setup `passthru.updateScript`.

Only build or run rke2 in linux.

mogeko committed:
nixos/rke2: add rke2 service

Set assertions to avoid obvious errors.

Eliminate the conflict between default CNI (`cana`) and `NetworkManager`.

Determine whether optional can be used for agent.

Add the option `cisHardening` to enable CIS Hardening.

Set kernel parameters by `boot.kernel.sysctl`.

Using `lib.escapeShellArgs` to make `ExecStart` more resilient to escaping issues.

Using a list of `str` to extra flags.

mogeko committed:
nixos/tests/rke2: add tests for single-node and multi-node

For multi-node test, create a 3-node cluster (2 server nodes and 1 agent node) to test the connection between nodes.

Setup `passthru.tests`.

mogeko commented:
> Commit: [nixos/rke2: complete the document](https://github.com/NixOS/nixpkgs/pull/308090/commits/819fe7eee572b319d457cd44cfd323e98f0d820c) Has CI failing. Please fix error. https://github.com/NixOS/nixpkgs/actions/runs/8910965374/job/24471302839

It has been fixed.

zimbatm commented:
Just noticed this PR. I didn't look into the code yet, but you might want to steal some of the code from https://github.com/numtide/nixos-rke2/blob/main/modules/nixos/rke2/default.nix 
This code is battle-tested in a few deployments already.

mogeko commented:
> Just noticed this PR. I didn't look into the code yet, but you might want to steal some of the code from https://github.com/numtide/nixos-rke2/blob/main/modules/nixos/rke2/default.nix This code is battle-tested in a few deployments already.

Thank you for your suggestion!

I have carefully studied [its code](https://github.com/numtide/nixos-rke2/blob/main/modules/nixos/rke2/default.nix ). **Our general idea is the same**, but the way it sets kernel parameters is more reasonable.

I have transplanted this part of the code.

zimbatm commented on nixos/modules/services/cluster/rke2/default.nix:
```diff
@@ -0,0 +1,317 @@
+    };
+
+    extraFlags = mkOption {
+      type = types.str;
```
Having this a list of str makes it more useful since then multiple modules can append to the list.

zimbatm commented on nixos/modules/services/cluster/rke2/default.nix:
```diff
@@ -0,0 +1,317 @@
+          "-${pkgs.kmod}/bin/modprobe br_netfilter"
+          "-${pkgs.kmod}/bin/modprobe overlay"
+        ];
+        ExecStart = concatStringsSep " \\\n  " (
```
Using lib.escapeShellArgs would make it more resilient to escaping issues

zimbatm reviewed (commented):
Finally had some time to review. Found mino nits, no blocker .

zimbatm reviewed (approved):
:lgtm:
